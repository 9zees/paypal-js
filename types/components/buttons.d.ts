import type { CreateOrderRequestBody, CaptureOrderResponseBody } from '../apis/orders';
import type { CreateSubscriptionRequestBody } from '../apis/subscriptions';

type UnknownObject = Record<string, unknown>;

export type CreateOrderActions = {
    order: {
        /** Used to create an order for client-side integrations. Accepts the same options as the request body of the [/v2/checkout/orders api](https://developer.paypal.com/docs/api/orders/v2/#orders-create-request-body). */
        create: (options: CreateOrderRequestBody) => Promise<string>;
    }
};

export type CreateSubscriptionActions = {
    subscription: {
        /** Used to create a subscription for client-side integrations. Accepts the same options as the request body of the [/v1/billing/subscription api](https://developer.paypal.com/docs/api/subscriptions/v1#subscriptions-create-request-body). */
        create: (options: CreateSubscriptionRequestBody) => Promise<string>;
    }
};

export type OnInitActions = {
    enable: () => void;
    disable: () => void;
};

export type OnClickActions = {
    reject: () => void;
    resolve: () => void;
};

// todo: what else is possible here?
export type OnApproveOrderActions = {
    order: {
        capture: () => Promise<CaptureOrderResponseBody>;
        redirect: (redirectURL: string) => void;
        restart: () => void;

    }
};

export type OnCancelledOrderData = {
    orderID: string;
};

export type OnCancelledOrderActions = {
    redirect: () => void;
};

export interface PayPalButtonsComponentProps {
    /**
     * Called on button click. Often used for [Braintree vault integrations](https://developers.braintreepayments.com/guides/paypal/vault/javascript/v3).
     */
    createBillingAgreement?: () => Promise<string>;
    /**
     * Called on button click. Supports creating an order with the [/v2/checkout/orders api](https://developer.paypal.com/docs/api/orders/v2/#orders_create).
     */
    createOrder?: (data: UnknownObject, actions: CreateOrderActions) => Promise<string>;
    /**
     * Called on button click. Supports creating a subscription with the [/v1/billing/subscription api](https://developer.paypal.com/docs/api/subscriptions/v1/#subscriptions_create).
     */
    createSubscription?: (data: UnknownObject, actions: CreateSubscriptionActions) => Promise<string>;
    /**
     * Used for defining a standalone button.
     * Learn more about [configuring the funding source for standalone buttons](https://developer.paypal.com/docs/business/checkout/configure-payments/standalone-buttons/#4-funding-sources).
     */
    fundingSource?: string;
    /**
     * Called when finalizing the transaction. Supports capturing an order with the [v2/checkout/orders/:order_id/capture api](https://developer.paypal.com/docs/api/orders/v2/#orders_capture).
     */
    onApprove?: (data: UnknownObject, actions: OnApproveOrderActions) => Promise<void>;
    /**
     * Called when the buyer cancels the transaction.
     * Often used to show the buyer a [cancellation page](https://developer.paypal.com/docs/business/checkout/add-capabilities/buyer-experience/#3-show-cancellation-page).
     */
    onCancel?: (data: UnknownObject, actions: OnCancelledOrderActions) => void;
    /**
     * Called when the button is clicked. Often used for [validation](https://developer.paypal.com/docs/checkout/integration-features/validation/).
     */
    onClick?: (data: UnknownObject, actions: OnClickActions) => Promise<void> | void;
    /**
     * Catch all for errors preventing buyer checkout.
     * Often used to show the buyer an [error page](https://developer.paypal.com/docs/checkout/integration-features/handle-errors/).
     */
    onError?: (err: UnknownObject) => void;
    /**
     * Called when the buttons are initialized. The component is initialized after the iframe has successfully loaded.
     */
    onInit?: (data: UnknownObject, actions: OnInitActions) => void;
    /**
     * Called when the buyer changes their shipping address on PayPal.
     */
    onShippingChange?: () => void;
    /**
     * [Styling options](https://developer.paypal.com/docs/business/checkout/reference/style-guide/#customize-the-payment-buttons) for customizing the button appearance.
     */
    style?: {
        color?: "gold" | "blue" | "silver" | "white" | "black";
        height?: number;
        label?: "paypal" | "checkout" | "buynow" | "pay" | "installment";
        layout?: "vertical" | "horizontal";
        shape?: "rect" | "pill";
        tagline?: boolean;
    };
}

export interface PayPalButtonsComponent {
    close: () => Promise<void>;
    isEligible: () => boolean;
    render: (container: HTMLElement | string) => Promise<void>;
}
