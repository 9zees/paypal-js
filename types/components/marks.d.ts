export interface PayPalMarksComponentProps {
    fundingSource?: string;
}

export interface PayPalMarksComponent {
    isEligible: () => boolean;
    render: (container: HTMLElement | string) => Promise<void>;
}
