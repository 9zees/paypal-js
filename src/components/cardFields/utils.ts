import { DEFAULT_PAYPAL_NAMESPACE, SDK_SETTINGS } from "../../constants";

import type {
    PayPalCardFieldsIndividualField,
    PayPalCardFieldsNamespace,
} from "../../types/payPalCardFieldsTypes";

/**
 * Throw an exception if the CardFields is not found in the paypal namespace
 * Probably cause for this problem is not sending the card-fields string
 * as part of the components props in options
 * {@code <PayPalScriptProvider options={{ components: 'card-fields'}}>}
 *
 * @param param0 and object containing the components and namespace defined in options
 * @throws {@code Error}
 *
 */
export const generateMissingCardFieldsError = ({
    components = "",
    [SDK_SETTINGS.DATA_NAMESPACE]: dataNamespace = DEFAULT_PAYPAL_NAMESPACE,
}: PayPalCardFieldsNamespace): string => {
    const expectedComponents = components
        ? `${components},card-fields`
        : "card-fields";
    let errorMessage = `Unable to render <PayPalCardFieldsProvider /> because window.${dataNamespace}.CardFields is undefined.`;

    if (!components.includes("card-fields")) {
        errorMessage += `\nTo fix the issue, add 'card-fields' to the list of components passed to the parent PayPalScriptProvider: <PayPalScriptProvider options={{ components: '${expectedComponents}'}}>`;
    }

    return errorMessage;
};

export function closeField(
    field: PayPalCardFieldsIndividualField | null,
    parentId: string
): void {
    const parentElement = document.getElementById(parentId);
    console.log({ parentElement, children: parentElement?.children });

    if (!parentElement?.hasChildNodes()) {
        console.log("element already closed");
        return;
    }
    if (!field) {
        console.log("no field!!");
    }
    console.log({ field });
    try {
        if (field && field.close) {
            field?.close();
        }
        // .catch((err) => {
        //     console.log("closeField Error: ", err);
        //     // noop
        // });
    } catch (error) {
        console.log("SEBSHRADDHA", { error });
    }
}
