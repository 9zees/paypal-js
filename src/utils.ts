import type { PayPalScriptOptions } from "../types/script-options";

interface StringMap {
    [key: string]: string;
}

export function findScript(
    url: string,
    dataAttributes?: StringMap
): HTMLScriptElement | null {
    const currentScript = document.querySelector<HTMLScriptElement>(
        `script[src="${url}"]`
    );
    if (currentScript === null) return null;

    const nextScript = createScriptElement(url, dataAttributes);

    // check if the new script has the same number of data attributes
    if (
        Object.keys(currentScript.dataset).length !==
        Object.keys(nextScript.dataset).length
    ) {
        return null;
    }

    let isExactMatch = true;

    // check if the data attribute values are the same
    Object.keys(currentScript.dataset).forEach((key) => {
        if (currentScript.dataset[key] !== nextScript.dataset[key]) {
            isExactMatch = false;
        }
    });

    return isExactMatch ? currentScript : null;
}

export interface ScriptElement {
    url: string;
    dataAttributes?: StringMap;
    onSuccess: () => void;
    onError: OnErrorEventHandler;
}

export function insertScriptElement({
    url,
    dataAttributes,
    onSuccess,
    onError,
}: ScriptElement): void {
    const newScript = createScriptElement(url, dataAttributes);
    newScript.onerror = onError;
    newScript.onload = onSuccess;

    document.head.insertBefore(newScript, document.head.firstElementChild);
}

export function processOptions(
    options: PayPalScriptOptions
): { url: string; dataAttributes: StringMap } {
    let sdkBaseURL = "https://www.paypal.com/sdk/js";

    if (options.sdkBaseURL) {
        sdkBaseURL = options.sdkBaseURL;
        delete options.sdkBaseURL;
    }

    interface ProcessedScriptOptions {
        queryParams: StringMap;
        dataAttributes: StringMap;
    }

    const processedOptions: ProcessedScriptOptions = {
        queryParams: {},
        dataAttributes: {},
    };

    Object.keys(options).forEach((key) => {
        const keyType =
            key.substring(0, 5) === "data-" ? "dataAttributes" : "queryParams";
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        processedOptions[keyType][key] = options[key]!.toString();
    });

    const { queryParams, dataAttributes } = processedOptions;

    return {
        url: `${sdkBaseURL}?${objectToQueryString(queryParams)}`,
        dataAttributes,
    };
}

export function objectToQueryString(params: StringMap): string {
    let queryString = "";

    Object.keys(params).forEach((key) => {
        if (queryString.length !== 0) queryString += "&";
        queryString += key + "=" + params[key];
    });
    return queryString;
}

function createScriptElement(
    url: string,
    dataAttributes: StringMap = {}
): HTMLScriptElement {
    const newScript: HTMLScriptElement = document.createElement("script");
    newScript.src = url;

    Object.keys(dataAttributes).forEach((key) => {
        newScript.setAttribute(key, dataAttributes[key]);

        if (key === "data-csp-nonce") {
            newScript.setAttribute("nonce", dataAttributes["data-csp-nonce"]);
        }
    });

    return newScript;
}
