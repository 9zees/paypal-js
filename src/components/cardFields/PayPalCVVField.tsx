import React, { useEffect, useRef, useState } from "react";

import { type PayPalCardFieldsIndividualFieldOptions } from "../../types";
import { usePayPalCardFields } from "./hooks";
import { ignore } from "./utils";

export const PayPalCVVField: React.FC<
    PayPalCardFieldsIndividualFieldOptions
> = (options) => {
    const { cardFields, cvvField } = usePayPalCardFields();

    const cvvContainer = useRef<HTMLDivElement>(null);
    // const cvvRef = useRef<PayPalCardFieldsIndividualField | null>(null);

    // We set the error inside state so that it can be caught by React's error boundary
    const [, setError] = useState(null);

    function closeComponent() {
        cvvField.current?.close().catch(ignore);
    }

    useEffect(() => {
        if (!cardFields.current || !cvvContainer.current) {
            return closeComponent;
        }

        cvvField.current = cardFields.current.CVVField(options);

        cvvField.current.render(cvvContainer.current).catch((err) => {
            // component failed to render, possibly because it was closed or destroyed.
            const cvvIsRendered = !!cvvContainer.current?.children.length;
            if (!cvvIsRendered) {
                // Component no longer in the DOM, we can safely ignore the error
                return;
            }
            // Component is still in the DOM
            setError(() => {
                throw new Error(
                    `Failed to render <PayPalCVVField /> component. ${err}`
                );
            });
        });

        return closeComponent;
    }, []); // eslint-disable-line react-hooks/exhaustive-deps

    return <div ref={cvvContainer} />;
};
