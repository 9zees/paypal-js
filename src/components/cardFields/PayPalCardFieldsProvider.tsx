import React, { ReactNode, useEffect, useRef, useState } from "react";

import { SCRIPT_LOADING_STATE } from "../../types";
import { useScriptProviderContext } from "../../hooks/scriptProviderHooks";
import { getPayPalWindowNamespace } from "../../utils";
import { SDK_SETTINGS } from "../../constants";
import { generateMissingCardFieldsError } from "./utils";
import { PayPalCardFieldsContext } from "./context";
import { usePayPalCardFieldsRegistry } from "./hooks";

import type {
    PayPalCardFieldsComponentOptions,
    PayPalCardFieldsComponent,
    PayPalCardFieldsIndividualField,
} from "@paypal/paypal-js/types/components/card-fields";

type CardFieldsProviderProps = PayPalCardFieldsComponentOptions & {
    children: ReactNode;
};

export const PayPalCardFieldsProvider = ({
    children,
    ...props
}: CardFieldsProviderProps): JSX.Element => {
    const [{ options, loadingStatus }] = useScriptProviderContext();
    const { registerField, unregisterField } = usePayPalCardFieldsRegistry();

    const cardFields = useRef<PayPalCardFieldsComponent | null>(null);

    const nameField = useRef<PayPalCardFieldsIndividualField | null>(null);
    const numberField = useRef<PayPalCardFieldsIndividualField | null>(null);
    const cvvField = useRef<PayPalCardFieldsIndividualField | null>(null);
    const expiryField = useRef<PayPalCardFieldsIndividualField | null>(null);

    const [isEligible, setIsEligible] = useState(false);
    // We set the error inside state so that it can be caught by React's error boundary
    const [, setError] = useState(null);

    useEffect(() => {
        // Only render the card fields when script is loaded and cardFields is eligible
        if (!(loadingStatus === SCRIPT_LOADING_STATE.RESOLVED)) {
            return;
        }

        try {
            cardFields.current =
                getPayPalWindowNamespace(
                    options[SDK_SETTINGS.DATA_NAMESPACE]
                ).CardFields?.({
                    ...props,
                }) ?? null;
        } catch (error) {
            setError(() => {
                throw new Error(
                    `Failed to render <PayPalCardFieldsProvider /> component. Failed to initialize:  ${error}`
                );
            });
            return;
        }

        if (!cardFields.current) {
            setError(() => {
                throw new Error(
                    generateMissingCardFieldsError({
                        components: options.components,
                        [SDK_SETTINGS.DATA_NAMESPACE]:
                            options[SDK_SETTINGS.DATA_NAMESPACE],
                    })
                );
            });
            return;
        }

        setIsEligible(cardFields.current.isEligible());

        // Clean up after component unmounts
        return () => {
            cardFields.current = null;
        };
    }, [loadingStatus]); // eslint-disable-line react-hooks/exhaustive-deps

    if (!isEligible) {
        // TODO: What should be returned here?
        return <div />;
    }

    return (
        <PayPalCardFieldsContext.Provider
            value={{
                cardFields,
                nameField,
                numberField,
                cvvField,
                expiryField,
                registerField,
                unregisterField,
            }}
        >
            {children}
        </PayPalCardFieldsContext.Provider>
    );
};
