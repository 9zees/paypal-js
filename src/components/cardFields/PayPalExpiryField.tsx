import React, { useEffect, useRef, useState } from "react";
import {
    PayPalCardFieldsIndividualField,
    PayPalCardFieldsIndividualFieldOptions,
} from "@paypal/paypal-js/types/components/card-fields";

import { usePayPalCardFields, usePayPalCardFieldsRenderState } from "./hooks";
import { hasChildren, ignore } from "./utils";

export const PayPalExpiryField: React.FC<
    PayPalCardFieldsIndividualFieldOptions
> = (options) => {
    const { cardFields, expiryField, expiryContainer } = usePayPalCardFields();
    const { registerField, unregisterField } = usePayPalCardFieldsRenderState();

    // const expiryContainer = useRef<HTMLDivElement>(null);
    // const expiryField = useRef<PayPalCardFieldsIndividualField | null>(null);

    // Set errors is state so that they can be caught by React's error boundary
    const [, setError] = useState(null);

    function closeComponent() {
        unregisterField("PayPalExpiryField");
        expiryField.current?.close().catch(ignore);
    }

    useEffect(() => {
        if (!cardFields.current || !expiryContainer.current) {
            return closeComponent;
        }

        registerField("PayPalExpiryField");
        expiryField.current = cardFields.current.ExpiryField(options);

        // Assigning current expiryField to the globally available expiryField in usePayPalCardFields() for merchants to interact with
        // globalExpiryField.current = expiryField.current;

        expiryField.current.render(expiryContainer.current).catch((err) => {
            if (!hasChildren(expiryContainer)) {
                // Component no longer in the DOM, we can safely ignore the error
                return;
            }
            // Component is still in the DOM
            setError(() => {
                throw new Error(
                    `Failed to render <PayPalExpiryField /> component. ${err}`
                );
            });
        });

        return closeComponent;
    }, []); // eslint-disable-line react-hooks/exhaustive-deps

    return <div ref={expiryContainer} />;
};
