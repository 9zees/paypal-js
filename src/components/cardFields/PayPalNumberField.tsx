import React, { useEffect, useRef, useState } from "react";

import { type PayPalCardFieldsIndividualFieldOptions } from "../../types";
import { usePayPalCardFields } from "./hooks";
import { ignore } from "./utils";

export const PayPalNumberField: React.FC<
    PayPalCardFieldsIndividualFieldOptions
> = (options) => {
    const { cardFields, numberField } = usePayPalCardFields();

    const numberContainer = useRef<HTMLDivElement>(null);

    // We set the error inside state so that it can be caught by React's error boundary
    const [, setError] = useState(null);

    function closeComponent() {
        numberField.current?.close().catch(ignore);
    }

    useEffect(() => {
        if (!cardFields.current || !numberContainer.current) {
            return closeComponent;
        }

        numberField.current = cardFields.current.NumberField(options);
        numberField.current.render(numberContainer.current).catch((err) => {
            const numberIsRendered = !!numberContainer.current?.children.length;
            if (!numberIsRendered) {
                // Component no longer in the DOM, we can safely ignore the error
                return;
            }
            // Component is still in the DOM
            setError(() => {
                throw new Error(
                    `Failed to render <PayPalNumberField /> component. ${err}`
                );
            });
        });

        return closeComponent;
    }, []); // eslint-disable-line react-hooks/exhaustive-deps

    return <div ref={numberContainer} />;
};
