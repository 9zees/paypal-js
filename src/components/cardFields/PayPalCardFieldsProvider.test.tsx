import React from "react";
import { render, waitFor } from "@testing-library/react";
import { ErrorBoundary } from "react-error-boundary";
import { loadScript } from "@paypal/paypal-js";
import { mock } from "jest-mock-extended";
import { PayPalCardFieldsComponent } from "@paypal/paypal-js/types/components/card-fields";

import { PayPalScriptProvider } from "../PayPalScriptProvider";
import { PayPalCardFieldsProvider } from "./PayPalCardFieldsProvider";
import {
    CARD_FIELDS_DUPLICATE_CHILDREN_ERROR,
    EMPTY_BRAINTREE_AUTHORIZATION_ERROR_MESSAGE,
} from "../../constants";
import { PayPalNumberField } from "./PayPalNumberField";
import { PayPalCVVField } from "./PayPalCVVField";
import { PayPalExpiryField } from "./PayPalExpiryField";
import { zoidCardFieldsComponents } from "./utils";

import type { ReactNode } from "react";

const mockCreateOrder = mock<() => Promise<string>>();
const mockOnApprove = mock<() => Promise<string>>();
const mockOnError = mock<() => void>();
const onError = jest.fn();
const wrapper = ({ children }: { children: ReactNode }) => (
    <ErrorBoundary fallback={<div>Error</div>} onError={onError}>
        {children}
    </ErrorBoundary>
);

jest.mock("@paypal/paypal-js", () => ({
    loadScript: jest.fn(),
}));

describe("PayPalCardFieldsProvider", () => {
    const isEligible = jest.fn();
    const mockRender = jest.fn();

    const componentMethods = {
        render: mockRender.mockResolvedValue({
            catch: jest.fn(),
        }),
        close: jest.fn().mockResolvedValue({
            catch: jest.fn(),
        }),
    };

    beforeEach(() => {
        document.body.innerHTML = "";

        window.paypal = {
            CardFields: () => {
                return {
                    isEligible: () => isEligible.mockReturnValue(true),
                    NumberField: () => componentMethods,
                    NameField: () => componentMethods,
                    CVVField: () => componentMethods,
                    ExpiryField: () => componentMethods,
                } as unknown as PayPalCardFieldsComponent;
            },
            version: "",
        };

        (loadScript as jest.Mock).mockResolvedValue(window.paypal);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    // PASS
    test("should throw an Error using the component without the PayPalScriptProvider", () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();

        render(
            <PayPalCardFieldsProvider
                onApprove={mockOnApprove}
                createOrder={mockCreateOrder}
                onError={mockOnError}
            >
                <></>
            </PayPalCardFieldsProvider>,
            { wrapper }
        );
        expect(onError.mock.calls[0][0].message).toEqual(
            "usePayPalScriptReducer must be used within a PayPalScriptProvider"
        );
        spyConsoleError.mockRestore();
    });

    // PASS
    test("should throw an Error using the component with PayPalScriptProvider without dataClientToken", () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();

        render(
            <PayPalScriptProvider options={{ clientId: "" }}>
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <></>
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>,
            { wrapper }
        );
        expect(onError.mock.calls[0][0].message).toEqual(
            EMPTY_BRAINTREE_AUTHORIZATION_ERROR_MESSAGE
        );
        spyConsoleError.mockRestore();
    });

    test("should throw an Error using the component with duplicate fields", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test-client",
                    currency: "USD",
                    intent: "authorize",
                    components: "card-fields",
                    dataClientToken: "test-data-client-token",
                }}
            >
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <PayPalNumberField />
                    <PayPalNumberField />
                    <PayPalCVVField />
                    <PayPalExpiryField />
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>,
            { wrapper }
        );
        await waitFor(() => {
            expect(onError.mock.calls[0][0].message).toEqual(
                CARD_FIELDS_DUPLICATE_CHILDREN_ERROR
            );
        });
        spyConsoleError.mockRestore();
    });

    // test("should return immediately when script provider is rejected", async () => {
    //     const spyConsoleError = jest
    //         .spyOn(console, "error")
    //         .mockImplementation();
    //     (loadScript as jest.Mock).mockRejectedValue(new Error("Unknown error"));

    //     render(
    //         <PayPalScriptProvider
    //             options={{
    //                 clientId: "test-client",
    //                 currency: "USD",
    //                 intent: "authorize",
    //                 components: "card-fields",
    //                 dataClientToken: "test-data-client-token",
    //             }}
    //         >
    //             <PayPalCardFieldsProvider
    //                 onApprove={mockOnApprove}
    //                 createOrder={mockCreateOrder}
    //                 onError={mockOnError}
    //             >
    //                 <PayPalNumberField />
    //                 <PayPalCVVField />
    //                 <PayPalExpiryField />
    //             </PayPalCardFieldsProvider>
    //         </PayPalScriptProvider>
    //     );
    //     await waitFor(() => {
    //         expect(loadScript).toBeCalled();
    //     });
    //     spyConsoleError.mockRestore();
    // });

    test.only("should not render CardFields components when unilegible", async () => {
        isEligible.mockReturnValue(false);

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test-client",
                    currency: "USD",
                    intent: "authorize",
                    components: "card-fields",
                    dataClientToken: "test-data-client-token",
                }}
            >
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <PayPalNumberField />
                    <PayPalCVVField />
                    <PayPalExpiryField />
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>
        );
        await waitFor(() => {
            expect(zoidCardFieldsComponents("number").length).toEqual(0);
            expect(zoidCardFieldsComponents("expiry").length).toEqual(0);
            expect(zoidCardFieldsComponents("cvv").length).toEqual(0);
        });
    });

    test("should throw an Error on hosted fields render process exception", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();

        mockRender.mockReturnValue({
            catch: jest.fn().mockRejectedValue(new Error("fail")),
        });

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test-client",
                    currency: "USD",
                    intent: "authorize",
                    dataClientToken: "test-data-client-token",
                }}
            >
                <PayPalCardFieldsProvider
                    onApprove={mockOnApprove}
                    createOrder={mockCreateOrder}
                    onError={mockOnError}
                >
                    <PayPalNumberField />
                    <PayPalCVVField />
                    <PayPalExpiryField />
                </PayPalCardFieldsProvider>
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => {
            expect(onError.mock.calls[0][0].message).toEqual(
                "Failed to render <PayPalCardFieldsProvider /> component. Error: Failing rendering CardFields"
            );
        });
        spyConsoleError.mockRestore();
    });

    // test("should render card fields", async () => {
    //     const { container, rerender } = render(
    //         <PayPalScriptProvider
    //             options={{
    //                 clientId: "test-client",
    //                 currency: "USD",
    //                 intent: "authorize",
    //                 dataClientToken: "test-data-client-token",
    //                 components: "card-fields",
    //             }}
    //         >
    //             <PayPalCardFieldsProvider createOrder={mockCreateOrder}>
    //                 <PayPalNumberField />
    //                 <PayPalCVVField />
    //                 <PayPalExpiryField />
    //             </PayPalCardFieldsProvider>
    //         </PayPalScriptProvider>
    //     );

    //     await waitFor(() => {
    //         expect(
    //             window?.paypal?.CardFields?.().NumberField()?.render
    //         ).toBeCalled();
    //     });
    //     expect(
    //         container.querySelector(".number") instanceof HTMLDivElement
    //     ).toBeTruthy();
    //     expect(
    //         container.querySelector(".expiration") instanceof HTMLDivElement
    //     ).toBeTruthy();
    //     expect(container.querySelector(".cvv")).toBeTruthy();

    //     // Rerender the component with new styles props
    //     // Shouldn't change the CardFields refs when rerendering
    //     rerender(
    //         <PayPalScriptProvider
    //             options={{
    //                 clientId: "test-client",
    //                 currency: "USD",
    //                 intent: "authorize",
    //                 dataClientToken: "test-data-client-token",
    //             }}
    //         >
    //             <PayPalCardFieldsProvider
    //                 createOrder={mockCreateOrder}
    //                 // styles={{
    //                 //     ".valid": { color: "#28a745" },
    //                 //     ".invalid": { color: "#dc3545" },
    //                 // }}
    //             >
    //                 <PayPalNumberField />
    //                 <PayPalCVVField />
    //                 <PayPalExpiryField />
    //             </PayPalCardFieldsProvider>
    //         </PayPalScriptProvider>
    //     );
    //     await waitFor(() => {
    //         expect(window?.paypal?.CardFields?.render).toBeCalledTimes(2);
    //     });
    // });
    // test("should not set context state if component is unmounted", async () => {
    //     jest.useFakeTimers();

    //     (
    //         (window.paypal as PayPalNamespace)
    //             .CardFields as PayPalCardFieldsComponent
    //     ).render = jest
    //         .fn()
    //         .mockImplementation(
    //             () =>
    //                 new Promise((resolve) => setTimeout(() => resolve({}), 500))
    //         );

    //     const { container, unmount } = render(
    //         <PayPalScriptProvider
    //             options={{
    //                 clientId: "test-client",
    //                 currency: "USD",
    //                 intent: "authorize",
    //                 dataClientToken: "test-data-client-token",
    //                 components: "card-fields",
    //             }}
    //         >
    //             <PayPalCardFieldsProvider createOrder={mockCreateOrder}>
    //                 <PayPalNumberField />
    //                 <PayPalCVVField />
    //                 <PayPalExpiryField />
    //             </PayPalCardFieldsProvider>
    //         </PayPalScriptProvider>
    //     );

    //     await waitFor(() => {
    //         expect(window?.paypal?.CardFields?.render).toBeCalled();
    //     });
    //     unmount();
    //     jest.runAllTimers();

    //     expect(
    //         container.querySelector(".number") instanceof HTMLDivElement
    //     ).toBeFalsy();
    //     jest.useRealTimers();
    // });
});
