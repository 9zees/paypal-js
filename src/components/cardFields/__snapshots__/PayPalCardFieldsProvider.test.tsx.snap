// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayPalCardFieldsProvider should throw an error when the 'card-fields' component is missing from the components list passed to the PayPalScriptProvider 1`] = `"Invalid authorization data. Use dataClientToken or dataUserIdToken to authorize."`;
